{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "779a4468",
   "metadata": {},
   "source": [
    "IPython\n",
    "=======\n",
    "\n",
    "\n",
    "Table of Contents\n",
    "-----------------\n",
    "\n",
    "* [Tips](#tips)\n",
    "   * [Tip 1: quickref](#tip-1-quickref)\n",
    "   * [Tip 2: Save history to file](#tip-2-save-history-to-file)\n",
    "* [Troubleshooting](#troubleshooting)\n",
    "   * [unexpected keyword argument 'column'](#unexpected-keyword-argument-column)\n",
    "   * [Unhandled exception in event loop](#unhandled-exception-in-event-loop)\n",
    "\n",
    "\n",
    "Tips\n",
    "----\n",
    "\n",
    "### Tip 1: quickref\n",
    "\n",
    "You can use the `%quickref` magic command to get a a quick introduction to\n",
    "IPythons features.\n",
    "\n",
    "```{code-block} text\n",
    "1>>> %quickref\n",
    "\n",
    "IPython -- An enhanced Interactive Python - Quick Reference Card\n",
    "================================================================\n",
    "\n",
    "obj?, obj??      : Get help, or more help for object (also works as\n",
    "                   ?obj, ??obj).\n",
    "?foo.*abc*       : List names in 'foo' containing 'abc' in them.\n",
    "%magic           : Information about IPython's 'magic' % functions.\n",
    "...\n",
    "```\n",
    "\n",
    "### Tip 2: Save history to file\n",
    "\n",
    "You can save the history from your last session to the file in the directory\n",
    "you were in when you started `ipython` by using the `%save` magic command.\n",
    "\n",
    "```{code-block} text\n",
    "---\n",
    "caption: ipython shell -- save to a file named session.py\n",
    "---\n",
    "1>>> %save session ~1/\n",
    "The following commands were written to file `session.py`:\n",
    "import requests\n",
    "response = requests.get(\"http://api.punkapi.com/v2/beers/random\")\n",
    "response.ok\n",
    "response.status_code\n",
    "response.reason\n",
    "...\n",
    "```\n",
    "\n",
    "Troubleshooting\n",
    "---------------\n",
    "\n",
    "### unexpected keyword argument 'column'\n",
    "\n",
    ":::{dropdown} Error\n",
    ":open:\n",
    ":title: card-title\n",
    "\n",
    "This error occurs when attempting tab completion.\n",
    "\n",
    "```\n",
    "  File \"../venv/lib/python3.8/site-packages/IPython/core/completer.py\", line 2029, in _complete\n",
    "    completions = self._jedi_matches(\n",
    "  File \"../venv/lib/python3.8/site-packages/IPython/core/completer.py\", line 1373, in _jedi_matches\n",
    "    interpreter = jedi.Interpreter(\n",
    "  File \"../venv/lib/python3.8/site-packages/jedi/api/__init__.py\", line 725, in __init__\n",
    "    super().__init__(code, environment=environment,\n",
    "TypeError: __init__() got an unexpected keyword argument 'column'\n",
    "```\n",
    "\n",
    ":::\n",
    "\n",
    ":::{dropdown} Solution\n",
    ":open:\n",
    "\n",
    "iPython `<=7.19` is incompatable with jedi version `0.18.0`.\n",
    "To resolve, downgrade `jedi` to version `0.17.2`.\n",
    "\n",
    "`````{tabbed} poetry\n",
    "\n",
    "Modify your {file}`pyproject.toml` file and add the following `jedi`\n",
    "dependency just above the `ipython` dependency.\n",
    "\n",
    "```{code-block} toml\n",
    ":caption: pyproject.toml\n",
    ":emphasize-lines: \"3-3\"\n",
    "\n",
    "[tool.poetry.dev-dependencies]\n",
    "...\n",
    "jedi = \"0.17.2\" # ipython <=7.19\n",
    "ipython = \"^7.19.0\"\n",
    "```\n",
    "\n",
    "Then in a terminal run:\n",
    "\n",
    "```{code-block} bash\n",
    ":caption: terminal\n",
    "poetry update jedi\n",
    "```\n",
    "\n",
    "`````\n",
    "\n",
    "`````{tabbed} pip\n",
    "\n",
    "In a terminal run:\n",
    "\n",
    "```{code-block} bash\n",
    ":caption: terminal\n",
    "python -m pip install -U jedi~=0.17.2\n",
    "```\n",
    "\n",
    "`````\n",
    "\n",
    ":::\n",
    "\n",
    "```{seealso}\n",
    "\n",
    "* [ipython/ipython#12740](https://github.com/ipython/ipython/issues/12740)\n",
    "\n",
    "```\n",
    "\n",
    "### Unhandled exception in event loop\n",
    "\n",
    ":::{dropdown} Error\n",
    ":open:\n",
    ":title: card-title\n",
    "\n",
    "This error occurs intermittently after hitting enter.\n",
    "\n",
    "```\n",
    "Unhandled exception in event loop:\n",
    "  File \"...\\lib\\asyncio\\proactor_events.py\", line 768, in _loop_self_reading\n",
    "\n",
    "    f.result()  # may raise\n",
    "  File \"...\\lib\\asyncio\\windows_events.py\", line 808, in _poll\n",
    "    value = callback(transferred, key, ov)\n",
    "  File \"...\\lib\\asyncio\\windows_events.py\", line 457, in finish_recv\n",
    "    raise ConnectionResetError(*exc.args)\n",
    "\n",
    "Exception [WinError 995] The I/O operation has been aborted because of either a thread exit or an application request\n",
    "```\n",
    "\n",
    ":::\n",
    "\n",
    ":::{dropdown} Solution\n",
    ":open:\n",
    "\n",
    "iPython ~7.13.0 on Windows does not work with a version 3+ of prompt-toolkit.\n",
    "To resolve, downgrade `prompt-toolkit` to version `2.x`.\n",
    "\n",
    "`````{tabbed} poetry\n",
    "\n",
    "Modify your {file}`pyproject.toml` file and add the following `prompt-toolkit`\n",
    "dependency just above the `ipython` dependency.\n",
    "\n",
    "```{code-block} toml\n",
    ":caption: pyproject.toml\n",
    ":emphasize-lines: \"3-3\"\n",
    "\n",
    "[tool.poetry.dev-dependencies]\n",
    "...\n",
    "prompt-toolkit = \"^2\" # ipython ~7.10\n",
    "ipython = \"^7.13.0\"\n",
    "```\n",
    "\n",
    "Then in a terminal run:\n",
    "\n",
    "```{code-block} bash\n",
    ":caption: terminal\n",
    "poetry update prompt-toolkit\n",
    "```\n",
    "\n",
    "`````\n",
    "\n",
    "`````{tabbed} pip\n",
    "\n",
    "In a terminal run:\n",
    "\n",
    "```{code-block} bash\n",
    ":caption: terminal\n",
    "python -m pip install -U prompt-toolkit~=2.0\n",
    "```\n",
    "\n",
    "`````\n",
    "\n",
    ":::\n",
    "\n",
    "```{seealso}\n",
    "\n",
    "* [ipython/ipython#12049](https://github.com/ipython/ipython/issues/12049)\n",
    "* [StackOverflow: Why do I get an 'Unhandled exception in event loop' error on ipython](https://stackoverflow.com/questions/59355904/why-do-i-get-an-unhandled-exception-in-event-loop-error-on-ipython)\n",
    "\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "md:myst",
   "text_representation": {
    "extension": ".md",
    "format_name": "myst"
   }
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "source_map": [
   11
  ]
 },
 "nbformat": 4,
 "nbformat_minor": 5
}